src/josh.rs

```diff
@@ -0,0 +1,121 @@
+use crate::config::JoshConfig;
+use crate::utils::run_command;
+use anyhow::Context;
+use std::net::{SocketAddr, TcpStream};
+use std::path::PathBuf;
+use std::process::{Command, Stdio};
+use std::time::Duration;
+
+const JOSH_PORT: u16 = 42042;
+
+pub struct JoshProxy {
+    path: PathBuf,
+}
+
+impl JoshProxy {
+    /// Tries to figure out if `josh-proxy` is installed.
+    pub fn lookup() -> Option<Self> {
+        which::which("josh-proxy").ok().map(|path| Self { path })
+    }
+
+    pub fn start(&self, config: &JoshConfig) -> anyhow::Result<RunningJoshProxy> {
+        // Determine cache directory.
+        let user_dirs =
+            directories::ProjectDirs::from("org", &config.full_repo_name(), "rustc-josh")
+                .context("cannot determine cache directory for Josh")?;
+        let local_dir = user_dirs.cache_dir().to_owned();
+
+        // Start josh, silencing its output.
+        let josh = std::process::Command::new(&self.path)
+            .arg("--local")
+            .arg(local_dir)
+            .args([
+                "--remote=https://github.com",
+                &format!("--port={JOSH_PORT}"),
+                "--no-background",
+            ])
+            .stdout(Stdio::null())
+            .stderr(Stdio::null())
+            .spawn()
+            .context("failed to start josh-proxy, make sure it is installed")?;
+
+        // Wait until the port is open. We try every 10ms until 1s passed.
+        for _ in 0..100 {
+            // This will generally fail immediately when the port is still closed.
+            let addr = SocketAddr::from(([127, 0, 0, 1], JOSH_PORT));
+            let josh_ready = TcpStream::connect_timeout(&addr, Duration::from_millis(1));
+
+            if josh_ready.is_ok() {
+                println!("josh up and running");
+                return Ok(RunningJoshProxy {
+                    process: josh,
+                    port: JOSH_PORT,
+                });
+            }
+
+            // Not ready yet.
+            std::thread::sleep(Duration::from_millis(10));
+        }
+        panic!("Even after waiting for 1s, josh-proxy is still not available.");
+    }
+}
+
+pub fn try_install_josh() -> Option<JoshProxy> {
+    run_command(&[
+        "cargo",
+        "install",
+        "--locked",
+        "--git",
+        "https://github.com/josh-project/josh",
+        "--tag",
+        "r24.10.04",
```

Comment:
Might be worth moving the josh version into a separate constant so it is clear and easy to change in the future?